nixos/modules/services/web-apps/nextcloud.nix

```diff
@@ -728,14 +767,34 @@ in {
               ${optionalString (c.dbport != null) "'dbport' => '${toString c.dbport}',"}
               ${optionalString (c.dbuser != null) "'dbuser' => '${c.dbuser}',"}
               ${optionalString (c.dbtableprefix != null) "'dbtableprefix' => '${toString c.dbtableprefix}',"}
-              ${optionalString (c.dbpassFile != null) "'dbpassword' => nix_read_secret('${c.dbpassFile}'),"}
+              ${optionalString (c.dbpassFile != null) ''
+                  dbpassword' => nix_read_file(
+                    ${c.dbpassFile},
+                     "Cannot start Nextcloud, dbpass file %s set by NixOS doesn't seem to "
+                    . "exist! Please make sure that the file exists and has appropriate "
+                    . "permissions for user & group 'nextcloud'!"
+                  ),
+                ''
+              }
               'dbtype' => '${c.dbtype}',
               'trusted_domains' => ${writePhpArrary ([ cfg.hostName ] ++ c.extraTrustedDomains)},
               'trusted_proxies' => ${writePhpArrary (c.trustedProxies)},
               ${optionalString (c.defaultPhoneRegion != null) "'default_phone_region' => '${c.defaultPhoneRegion}',"}
               ${optionalString (nextcloudGreaterOrEqualThan "23") "'profile.enabled' => ${boolToString cfg.globalProfiles},"}
               ${objectstoreConfig}
             ];
+
+            $CONFIG = array_replace_recursive($CONFIG, nix_decode_json_file(
+              "${jsonFormat.generate "nextcloud-extraOptions.json" cfg.extraOptions}",
+              "decoding json extra options file failed"
```

Comment:
maybe?

I guess we could do this:

```nix
${ let extraConfig = jsonFormat.generate "nextcloud-extraOptions.json" cfg.extraOptions;
   in ''
     $CONFIG = array_replace_recursive($CONFIG, nix_decode_json_file(
     "${extraConfig}",
     "decoding the json-encoded file generated from the extraOptions failed (${extraConfig}). This is probably a bug in nixpkgs."
   ));
 ''}
```

Reply:
ah — tbh i don't like writing php (never properly learnt it, and every time i touch anything things seem to go up in flames), so i tend to do as much as possible on the nix side of things. But if you write a snippet / patch for that I'll be happy to add it in

Reply:
> This doesn't have %s although it can theoretically never fail. Maybe it makes more sense to just pass in "file type" such as "extraOptions file" or "secretFile file" and the function itself can generate a message and include the path?

AFAIU this is nothing an admin can break while deploying bogus configuration, but only if someone contributes code that's wrong. So I'd argue it's good enough and other failures just shouldn't pass reviews.

Reply:
I don't like passing in an invalid format string because it can't be used. Let's either pass in something trivial like `"this should never happen %s"` or handle this properly. Honestly i don't think we need to do anything too fancy, but I don't want to merge in "wrong" code as it will confuse future readers.

Reply:
I've changed it